testdata/generate.py

```diff
@@ -0,0 +1,526 @@
+        super(AllocationGenerator, self).__init__(conn, logger, distribution)
+        self.gau_count = gau_count
+
+    def __enter__(self):
+        super(AllocationGenerator, self).__enter__()
+
+        # generate GAUs
+        self.gaus = []
+        with GAUGenerator(self.conn, self.logger) as gau_gen:
+            for x in range(self.gau_count):
+                if x == 0:
+                    kw = {"name": "General Fund"}
+                else:
+                    kw = {}
+                self.gaus.append(gau_gen.make_one(**kw))
+
+        return self
+
+    def make_one(self, **kw):
+        attrs = {}
+        attrs["id"] = self.make_id()
+        attrs["percent"] = 100
+        attrs["gau_id"] = random.choice(self.gaus)["id"]
+        attrs.update(kw)
+        self.writer.writerow(attrs)
+        return attrs
+
+
+class OpportunityGenerator(Generator):
+    table = "opportunities"
+    columns = (
+        "id",
+        "name",
+        "amount",
+        "stage_name",
+        "close_date",
+        "dont_create_payments",
+        "account_id",
+        "contact_id",
+        "recurring_donation_id",
+        "campaign_id",
+    )
+
+    def init_subgenerators(self):
+        self.generators = {
+            "oppconroles": OpportunityContactRoleGenerator(self.conn, self.logger),
+            "payments": PaymentGenerator(
+                self.conn, self.logger, {1: 200, 2: 1, 3: 1, 4: 1}
+            ),
+            "partialsoftcredits": PartialSoftCreditGenerator(
+                self.conn, self.logger, {0: 200, 1: 1}
+            ),
+            "allocations": AllocationGenerator(self.conn, self.logger, {1: 9, 2: 1}),
+        }
+
+    def make_one(self, contacts, **kw):
+        attrs = {}
+        attrs["id"] = id = self.make_id()
+        attrs["amount"] = amount = random.randrange(100, 1100, 100)
+        close_date = fake.date_between_dates(ten_years_ago, today)
+        attrs["close_date"] = close_date.isoformat()
+        attrs["stage_name"] = "Closed Won"
+        attrs["dont_create_payments"] = False
+        attrs["recurring_donation_id"] = None
+        attrs["campaign_id"] = None
+        attrs.update(kw)
+        contact = contacts[0]
+        attrs["name"] = "{} {} Donation ({})".format(
+            contact["first_name"], contact["last_name"], attrs["close_date"]
+        )
+
+        # opportunity contact roles
+        oppconrole_gen = self.generators["oppconroles"]
+        first = True
+        for contact in contacts:
+            oppconrole_gen.make_one(
+                opportunity_id=id,
+                contact_id=contact["id"],
+                role="Donor" if first else "Influencer",
+                is_primary=first,
+            )
+            first = False
+
+        # payments
+        payment_gen = self.generators["payments"]
+        payment_count = payment_gen.choose_count()
+        payment_amount = round(amount / payment_count, 2)
+        for n in range(payment_count):
```

Comment:
Does it matter that `range()` has different behavior in python 2/3? Probably not, right?

Reply:
It's a small list, so I doubt it has much impact.